coil-base/src/main/java/coil/decode/ImageSource.kt

```diff
@@ -133,14 +141,18 @@ sealed class ImageSource : Closeable {
      * Return a [File] containing this [ImageSource]'s data.
      * If this image source is backed by a [BufferedSource], a temporary file containing this
      * [ImageSource]'s data will be created.
+     *
+     * This method is only safe when used on FileSystem.SYSTEM.
      */
-    abstract fun file(): File
+    abstract fun file(): Path
 
     /**
      * Return a [File] containing this [ImageSource]'s data if one has already been created.
      * Else, return 'null'.
      */
-    abstract fun fileOrNull(): File?
+    abstract fun fileOrNull(): Path?
+
+    abstract fun mediaDataSource(): MediaDataSource
```

Comment:
I don't think we should expose a custom `MediaDataSource` here. Originally `ImageSource` was only a `BufferedSource` to keep `Decoder`s agnostic to the data source. Unfortunately, Android has several decoders that are coupled with `File`s which is why `file` and `fileOrNull` were added. `ImageSource.Metadata` was added recently, however it's optional.